### ISSUE TITLE
[CFE] number parsing issue. 

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/sdk#54539]
Consider:

```dart
void main() {
  final a = [
    if (true) 0else 1,
  ];
}
```

I've expected that program to be a valid Dart program.

However, that program is rejected by the implementation with the following error messages:

```
Error compiling to JavaScript:
compile: lib/main.dart:3:15:
Error: Numbers in exponential notation should always contain an exponent (an integer number with an optional sign).
    if (true) 0else 1,
              ^
lib/main.dart:3:17:
Error: Expected 'else' or comma.
    if (true) 0else 1,
                ^^^
lib/main.dart:3:21:
Error: Expected ',' before this.
    if (true) 0else 1,
                    ^
lib/main.dart:3:17:
Error: Undefined name 'lse'.
    if (true) 0else 1,
                ^^^
Error: Compilation failed.
```

It appears to me that there's a bug in the scanner.

**Comment by lrhn** [comment]
I wouldn't expect it to be valid.
It might be, but I wouldn't _expect_ it by looking at the code.

Does seem that the `0e` is scanned as a double numeral, then rejected because it's incomplete, leaving the `lse` to be scanned by itself.
That's not consistent with the story that tokenization matches the longest valid token at a position, when `0e` is not actually a valid token in the lexical grammar.

**Comment by codesculpture** [comment]
I dont think this is a bug. This syntax is not acceptable.

**Comment by modulovalue** [comment]
> I dont think this is a bug. This syntax is not acceptable.

If we produce an NFA for a reduced lexical grammar that contains only those tokens that are relevant to this issue (i.e., else-keywords and integer-literals), and we determinize that NFA, then we get the following DFA:

<img width="1265" alt="Screenshot 2024-01-08 at 22 39 24" src="https://github.com/dart-lang/sdk/assets/24444584/29b679cd-cbd2-4106-88f5-38cf94f37205">


Given an input such as `0e•l`, on receiving an `l`, we enter an error state and, given that the spec specifies maximal munch lexing for all tokens, we have to accept the last accepting state, which is an integer literal `0•`. Afterwards, we have to start lexing from `•el` (that is, we have to jump back after the last accepting position), which will lead us to the else keyword after `else•`.

We will have lexed an integer literal `0` and an else keyword `else`.

If my understanding is correct, this is either a bug in the implementation or the spec needs to be adjusted to be explicit about the fact that integer literals do not depend on maximal munch but implement a different rule.

Since the latter appears to me to be quite uncommon, I think this should be considered a bug and the scanner should be adjusted to implement the behavior as specified.

**Comment by lrhn** [comment]
Agree. The [code](https://github.com/dart-lang/sdk/blob/ef75007c2f254a6b40872299da4db4d7b4079e5d/pkg/_fe_analyzer_shared/lib/src/scanner/abstract_scanner.dart#L1349) has a recovery when seeing `0e` not followed by `+`, `-` or a digit, instead of falling back to before the `e`.
Which is probably correct most of the time, since nobody wants to write `0else` in practice.
Still, it's not what the lexical grammar says.

It was not an issue until we introduced collection-`if`. Prior to that, an `else` had to be following a statement, and could therefore not follow a numeral. Statements (and declarations) end in either `}` or `;`.
With collection-`if`, you can now write `[if(b)0else-1]`, and almost get away with it!

**Comment by codesculpture** [comment]
> > I dont think this is a bug. This syntax is not acceptable.
> 
> If we produce an NFA for a reduced lexical grammar that contains only those tokens that are relevant to this issue (i.e., else-keywords and integer-literals), and we determinize that NFA, then we get the following DFA:
> 
> <img width="1265" alt="Screenshot 2024-01-08 at 22 39 24" src="https://github.com/dart-lang/sdk/assets/24444584/29b679cd-cbd2-4106-88f5-38cf94f37205">
> 
> 
> Given an input such as `0e•l`, on receiving an `l`, we enter an error state and, given that the spec specifies maximal munch lexing for all tokens, we have to accept the last accepting state, which is an integer literal `0•`. Afterwards, we have to start lexing from `•el` (that is, we have to jump back after the last accepting position), which will lead us to the else keyword after `else•`.
> 
> We will have lexed an integer literal `0` and an else keyword `else`.
> 
> If my understanding is correct, this is either a bug in the implementation or the spec needs to be adjusted to be explicit about the fact that integer literals do not depend on maximal munch but implement a different rule.
> 
> Since the latter appears to me to be quite uncommon, I think this should be considered a bug and the scanner should be adjusted to implement the behavior as specified.

You are right, but i think `[0-9]•else|if` grammar is only suits on inside collections [...]

**Comment by codesculpture** [comment]
```
[if (10 > 12) 0e23else 1]
```

This is accepted by dart
